### PR TITLE
Update unit test guideline for unimplemented test

### PR DESCRIPTION
**Pull request by kevindiu**
### Description:

This PR update the unit test guideline for the not-implemented placeholder implemented in https://github.com/vdaas/vald/pull/1967.

*grammar checked*

### Versions:

- Go Version: 1.20.2
- Docker Version: 20.10.8
- Kubernetes Version: 1.22.0
- NGT Version: 2.0.9

### Checklist:

- [x] I have read the [CONTRIBUTING](https://github.com/vdaas/vald/blob/main/CONTRIBUTING.md) document and completed [our CLA agreement](https://cla-assistant.io/vdaas/vald).
- [x] I have checked open [Pull Requests](https://github.com/vdaas/vald/pulls) for the similar feature or fixes?

### CODE REVIEW

File: docs/contributing/unit-test-guideline.md [hunk 1]
```diff
@@ -184,14 +184,19 @@ One of the features is that Go will convert a single value to a slice value when
...
+
 So, we define the basic unit case slightly different from [the basic test case](#Basic).
 
 This change is very clear and you can apply it easily.
 Our basic test case depends on the type of two variadic arguments.
 
 1. When input is `...interface{}`
    - We have to write all test cases which satisfy `...interface{}` as same as a [basic test case](#Basic). For example, `val = 1`, `val = "input"`, `val = []float64{2020.12}` and so on.
-1. When input is not `...interface{}` but `...[]int`, `...[]string` or etc.
+2. When input is not `...interface{}` but `...[]int`, `...[]string` or etc.
```
Comment (kevindiu) [inline docs/contributing/unit-test-guideline.md:199]:
```suggestion
1. When input is not `...interface{}` but `...[]int`, `...[]string` or etc.
```

File: docs/contributing/unit-test-guideline.md [hunk 2]
```diff
@@ -184,14 +184,15 @@ One of the features is that Go will convert a single value to a slice value when
 
 And we apply the table-driven test for running unit tests.
 For example, when we create the unit test of `func getMeta(...[]int)`, the test code will be more complex than other functions' tests which don't use variadic argument as the input, if we create the test for all input patterns.
+
```
Comment (kevindiu) [inline docs/contributing/unit-test-guideline.md:187]:
```suggestion
```

File: docs/contributing/unit-test-guideline.md [hunk 3]
```diff
@@ -184,14 +184,15 @@ One of the features is that Go will convert a single value to a slice value when
...
 So, we define the basic unit case slightly different from [the basic test case](#Basic).
 
 This change is very clear and you can apply it easily.
 Our basic test case depends on the type of two variadic arguments.
 
 1. When input is `...interface{}`
    - We have to write all test cases which satisfy `...interface{}` as same as a [basic test case](#Basic). For example, `val = 1`, `val = "input"`, `val = []float64{2020.12}` and so on.
-1. When input is not `...interface{}` but `...[]int`, `...[]string` or etc.
+2. When input is not `...interface{}` but `...[]int`, `...[]string` or etc.
```
Comment (kevindiu) [inline docs/contributing/unit-test-guideline.md:195]:
```suggestion
1. When input is not `...interface{}` but `...[]int`, `...[]string` or etc.
```

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -6,15 +6,18 @@ We appreciate your help!
...
+  - [Contributing Issue](#contributing-issue)
+    - [Bug Report](#bug-report)
+    - [Proposal](#proposal)
+    - [Feature Request](#feature-request)
+    - [Security Issue Report](#security-issue-report)
+  - [Contributing Source Code](#contributing-source-code)
+    - [Before contributing source code](#before-contributing-source-code)
+    - [How to contribute source code](#how-to-contribute-source-code)
+    - [How to contribute test code](#how-to-contribute-test-code)
+    - [Branch naming convention](#branch-naming-convention)
```
Comment (kevindiu) [inline CONTRIBUTING.md:20]:
```suggestion
- [Table of Contents](#table-of-contents)
- [Contributing Issue](#contributing-issue)
  - [Bug Report](#bug-report)
  - [Proposal](#proposal)
  - [Feature Request](#feature-request)
  - [Security Issue Report](#security-issue-report)
- [Contributing Source Code](#contributing-source-code)
  - [Before contributing source code](#before-contributing-source-code)
  - [How to contribute source code](#how-to-contribute-source-code)
  - [How to contribute test code](#how-to-contribute-test-code)
  - [Branch naming convention](#branch-naming-convention)
```

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -6,15 +6,17 @@ We appreciate your help!
...
-- [Contributing Issue](#Contributing-Issue)
-  - [Bug Report](#Bug-Report)
-  - [Proposal](#Proposal)
-  - [Feature Request](#Feature-Request)
-  - [Security Issue Report](#Security-Issue-Report)
-- [Contributing Source Code](#Contributing-Source-Code)
-  - [Before contributing source code](#Before-contributing-source-code)
-  - [How to contribute source code](#How-to-contribute-source-code)
-  - [Branch naming convention](#Branch-naming-convention)
+- [Table of Contents](#table-of-contents)
```
Comment (kevindiu) [inline CONTRIBUTING.md:9]:
```suggestion
```

Comment (kevindiu) [inline CONTRIBUTING.md:9]:
```suggestion
```

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -111,6 +112,85 @@ Your pull request is much more likely to be accepted if:
 
 - Your pull request is small and focused with a clear message that conveys the intent of your change
 
+### How to contribute test code
+
+To contribute test code, the steps are almost the same as contribute source code.
+However in Vald, we need some extra steps when making changes to test code.
```
Comment (kevindiu) [inline CONTRIBUTING.md:118]:
```suggestion
However, in Vald, we need some extra steps when making changes to test code.
```

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -111,6 +113,85 @@ Your pull request is much more likely to be accepted if:
...
+// NOT IMPLEMENTED BELOW
+
+func Test_notImplementedTest(t *testing.T) {
+    // this unit test function is not implemented yet
+}
+```
+
+When the test template is updated, if the unit test function is implemented already, we cannot do anything about it. But for unit test functions that are not implemented yet, we can re-generate them by using the updated template.
+
+Every unit test file should contains the placeholder in the file.
```
[the remark this answers is not in the document]
Comment (kevindiu) [reply CONTRIBUTING.md:142]:
```suggestion
Every unit test file should contain the placeholder in the file.
```

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -111,6 +113,85 @@ Your pull request is much more likely to be accepted if:
...
+If all of the test functions are implemented, the placeholder will be placed on the bottom of the unit test file.
```
[the remark this answers is not in the document]
Comment (kevindiu) [reply CONTRIBUTING.md:143]:
```suggestion
If all the test functions are implemented, the placeholder will be placed on the bottom of the unit test file.
```

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -111,6 +112,85 @@ Your pull request is much more likely to be accepted if:
 
 - Your pull request is small and focused with a clear message that conveys the intent of your change
 
+### How to contribute test code
+
+To contribute test code, the steps are almost the same as contribute source code.
+However, in Vald, we need some extra steps when making changes to test code.
```
Comment (vankichi) [inline CONTRIBUTING.md:118]:
```suggestion
But, there are some extra steps when making changes to the test code.
```

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -111,6 +112,85 @@ Your pull request is much more likely to be accepted if:
...
+
+In Vald, the table-driven test functions of the unit test are generated by [gotests](https://github.com/cweill/gotests) using the `make gotests/gen` command.
```
Comment (vankichi) [inline CONTRIBUTING.md:120]:
```suggestion
Vald applies the table-driven test functions of the unit test generated by [gotests](https://github.com/cweill/gotests) using the `make gotests/gen` command.
```

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -111,6 +112,85 @@ Your pull request is much more likely to be accepted if:
...
+The same style of implementation is generated for each unit test function, so everyone can follow the same style of implementing unit tests easily.
+
+We implemented our own [test template for gotests](https://github.com/vdaas/vald/tree/main/assets/test/templates) to generate the unit test functions, to provide extra flexibility and functionality for the unit test, such as [goleak](https://github.com/uber-go/goleak) support.
```
Comment (vankichi) [inline CONTRIBUTING.md:123]:
```suggestion
Vald uses its own [test template for gotests](https://github.com/vdaas/vald/tree/main/assets/test/templates) to generate the unit test functions, to provide extra flexibility and functionality for the unit test, such as [goleak](https://github.com/uber-go/goleak) support.
```

File: CONTRIBUTING.md [hunk 12]
```diff
@@ -111,6 +112,85 @@ Your pull request is much more likely to be accepted if:
...
+
+However, when we update our test template, the generated unit test functions will not be updated accordingly.
```
Comment (vankichi) [inline CONTRIBUTING.md:125]:
```suggestion
However, when the test template is updated, the generated unit test functions will not be updated accordingly.
```

File: CONTRIBUTING.md [hunk 13]
```diff
@@ -111,6 +112,85 @@ Your pull request is much more likely to be accepted if:
...
+To overcome this, we implemented a placeholder `// NOT IMPLEMENTED BELOW` to indicate whether the test functions are implemented or not. The unit test functions above the placeholder are implemented, and the functions below are not implemented yet.
```
Comment (vankichi) [inline CONTRIBUTING.md:126]:
```suggestion
To overcome this, each test file has a placeholder `// NOT IMPLEMENTED BELOW` to indicate whether the test functions are implemented.
The unit test functions above the placeholder are implemented, and the functions below still need to be implemented.
If all the test functions are implemented, the placeholder will be placed at the bottom of the unit test file.
```

File: CONTRIBUTING.md [hunk 14]
```diff
@@ -111,6 +112,85 @@ Your pull request is much more likely to be accepted if:
...
+}
+
+// NOT IMPLEMENTED BELOW
+
+func Test_notImplementedTest(t *testing.T) {
+    // this unit test function is not implemented yet
+}
+```
+
+When the test template is updated, if the unit test function is implemented already, we cannot do anything about it. But for unit test functions that are not implemented yet, we can re-generate them by using the updated template.
```
Comment (vankichi) [inline CONTRIBUTING.md:140]:
```suggestion
The implemented unit test is not updated even when the test template is updated.
On the other hand, we have to re-generate unimplemented tests by using the latest test template.
```

File: CONTRIBUTING.md [hunk 15]
```diff
@@ -111,6 +112,85 @@ Your pull request is much more likely to be accepted if:
...
+
+Every unit test file should contain the placeholder in the file.
+If all the test functions are implemented, the placeholder will be placed on the bottom of the unit test file.
```
Comment (vankichi) [inline CONTRIBUTING.md:143]:
```suggestion
```

**Review by vankichi — COMMENTED** [review]
You have to re-construct this chapter.
It seems to hard to understand what contributors should do when contributing tests.

This document requires the following:
1. How to contribute test code
    - not only a sentence. Please add all the steps when contributing test codes.
2. What is the structure of the test file

IMO, this PR does not satisfy all of the above yet and should be more systematic.

File: CONTRIBUTING.md [hunk 16]
```diff
@@ -111,6 +112,84 @@ Your pull request is much more likely to be accepted if:
...
+
+func Test_unimplementedTest(t *testing.T) {
+    // this unit test function is not implemented yet
+}
+```
+
+This implementation structure is generated by the `make gotests/gen` command for all unit test files, by appling the custom unit test template using [gotests](https://github.com/cweill/gotests).
+
+When the test template is updated, the generated unit test functions will not be updated accordingly, so we need to have this unit test structure to help us to regenerate the implementation with the latest template.
+For the implemented unit test, we cannot deal with it with the `make gotests/gen` command.
```
Comment (vankichi) [inline CONTRIBUTING.md:191]:
```suggestion
<div class="notice">
This implementation structure is generated by the <code class="bash" data-lang="bash>make gotests/gen</code> command for all unit test files by applying the custom unit test template using <a href="https://github.com/cweill/gotests">gotests</a>.
<br>
When the test template is updated, the generated unit test functions will not be updated accordingly, so we need this unit test structure to help us regenerate the implementation with the latest template.
For the implemented unit test, we cannot deal with it by executing  <code class="bash" data-lang="bash>make gotests/gen</code> command.
But we don't have to update implemented test soon.
</div>
```

**Review by vankichi — APPROVED** [review]

File: CONTRIBUTING.md [hunk 17]
```diff
@@ -111,6 +112,72 @@ Your pull request is much more likely to be accepted if:
...
+
+func Test_implementedTest2(t *testing.T) {
+    // this unit test is implemented
+}
+
+// NOT IMPLEMENTED BELOW
+```
+
+<div class="notice">
+This implementation structure is generated by the <code class="bash" data-lang="bash>make gotests/gen</code> command for all unit test files by applying the custom unit test template using <a href="https://github.com/cweill/gotests">gotests</a>.
```
Comment (vankichi) [inline CONTRIBUTING.md:174]:
```suggestion
This implementation structure is generated by the <code class="bash" data-lang="bash">make gotests/gen</code> command for all unit test files by applying the custom unit test template using <a href="https://github.com/cweill/gotests">gotests</a>.
```

File: CONTRIBUTING.md [hunk 18]
```diff
@@ -111,6 +112,72 @@ Your pull request is much more likely to be accepted if:
...
+<br>
+When the test template is updated, the generated unit test functions will not be updated accordingly, so we need this unit test structure to help us regenerate the implementation with the latest template.
+For the implemented unit test, we cannot deal with it by executing  <code class="bash" data-lang="bash>make gotests/gen</code> command.
```
Comment (vankichi) [inline CONTRIBUTING.md:177]:
```suggestion
For the implemented unit test, we cannot deal with it by executing  <code class="bash" data-lang="bash">make gotests/gen</code> command.
```

**Review by vankichi — CHANGES_REQUESTED** [review]

**Review by vankichi — APPROVED** [review]

**Review by datelier — APPROVED** [review]

**Pull request merged**